### PR TITLE
Fal 12 2 impl create user

### PR DESCRIPTION
**Pull request by Jomm0**
- Controller created for api, create endpoint
- User service created and create user implemented with exception handling
- validation handler class
- date and password utils
- all unit tested

### CODE REVIEW

File: user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java
Comment (willpennell) [inline user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java:1]: Looks good, would @Autowired be better to explicitly state which parts are DI'ed within the project.

File: user/src/test/java/com/jumar/user/services/impl/UserServiceImplTest.java
Comment (willpennell) [inline user/src/test/java/com/jumar/user/services/impl/UserServiceImplTest.java:1]: Looks good, change hardcoded test data, either to separate file or properties.

File: user/src/test/java/com/jumar/user/utils/DateUtilsTest.java
Comment (willpennell) [inline user/src/test/java/com/jumar/user/utils/DateUtilsTest.java:1]: Looks good, same as of test file, remove hardcoded test data

**Review by willpennell — APPROVED** [review]

File: user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java
Comment (Jomm0) [reply user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java:1]: As discussed, using Lombok to manage DI using private final access modifiers and @RequiredArgsConstructor annotations.

File: user/src/test/java/com/jumar/user/utils/DateUtilsTest.java
Comment (Jomm0) [reply user/src/test/java/com/jumar/user/utils/DateUtilsTest.java:1]: Will refactor test cases as discussed, in next PR.

File: user/src/main/java/com/jumar/user/dto/CreateUserDto.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.jumar.user.dto;
+
+import lombok.Data;
+
+import java.time.LocalDateTime;
+
+@Data
+public class CreateUserDto {
+    private String firstName;
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/dto/CreateUserDto.java:9]: We often use forenames as opposed to "first name"

File: user/src/main/java/com/jumar/user/dto/CreateUserDto.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    private String lastName;
+    private String emailAddress;
+    private String telephone;
+    private LocalDateTime dateOfBirth;
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/dto/CreateUserDto.java:13]: LocalDate - we won't have the time :)

File: user/src/main/java/com/jumar/user/exceptions/OneOrMoreUserRequiredFieldsNullException.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.jumar.user.exceptions;
+
+import java.io.Serial;
+
+public class OneOrMoreUserRequiredFieldsNullException extends IllegalArgumentException{
+    @Serial
+    private static final long serialVersionUID = 2;
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/exceptions/OneOrMoreUserRequiredFieldsNullException.java:7]: we don't normally bother about the serialVersionUID unless it's a project standard / convention. it's just so much noise

File: user/src/main/java/com/jumar/user/exceptions/validation/UserValidator.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import com.jumar.user.dto.CreateUserDto;
+import org.springframework.validation.Errors;
+import org.springframework.validation.Validator;
+
+public final class UserValidator {
+    private UserValidator(){}
+
+    public static boolean dtoFieldsAreNull(CreateUserDto createUserDto) {
+        return createUserDto.getFirstName() == null ||
+               createUserDto.getLastName() == null ||
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/exceptions/validation/UserValidator.java:12]: look into javax validation and NotNull annotations. You will need an @Valid tag in the controller method for it to work I believe.

File: user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+        ModelMapper mapper = new ModelMapper();
+
+        User newUser = mapper.map(createUserDto, User.class);
+        newUser.setDateAdded(LocalDateTime.now());
+        newUser.setDateLastModified(newUser.getDateAdded());
+        newUser.setUsername(newUser.getEmailAddress());
+        try{
+            newUser.setPasswordHash(PasswordUtils.hashPassword(newUser.getPasswordHash()));
+        } catch (NoSuchAlgorithmException e) {
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java:40]: what happens if you fail to hash it? the error should bubble up. don't catch it and print it out - this is called smothering exceptions. it's as if it never happened. if it's a checked exception put it in the throws section

File: user/src/main/java/com/jumar/user/utils/PasswordUtils.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    public static String hashPassword(String password) throws NoSuchAlgorithmException {
+        byte[] salt = saltGenerator();
+        MessageDigest md = MessageDigest.getInstance("SHA-512");
+
+        md.update(salt);
+
+        byte [] hashBytes = md.digest(password.getBytes(StandardCharsets.UTF_8));
+
+        return DatatypeConverter.printHexBinary(hashBytes).toUpperCase();
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/utils/PasswordUtils.java:20]: is there a Hash / Digest utils class that can do this?

File: user/src/main/java/com/jumar/user/utils/DateUtils.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.jumar.user.utils;
+
+import java.time.LocalDateTime;
+
+public final class DateUtils {
+    private DateUtils() {}
+
+    public static LocalDateTime dateOfBirthFormatter(String dateOfBirth) {
+
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/utils/DateUtils.java:9]: don't need this as 1) it's a LocalDate not LocalDateTime. 2) just parse it in the code or use an external library for dates.

File: user/src/main/java/com/jumar/user/utils/PasswordUtils.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    private static byte[] saltGenerator() {
+
+        return "supersecretsalt".getBytes();
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/utils/PasswordUtils.java:25]: this should be in the properties file

File: user/src/test/java/com/jumar/user/utils/DateUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import java.sql.Date;
+import java.time.LocalDateTime;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class DateUtilsTest {
+    private String dateOfBirth;
+
+    @Test
+    void converted_dateOfBirth_matches_correct_LocalDateTime_format() {
```
Comment (VincentFleetwood) [inline user/src/test/java/com/jumar/user/utils/DateUtilsTest.java:17]: this is testing the LocalDateTime class more than anything else - not business logic. We want our tests to test business logic.

File: user/src/test/java/com/jumar/user/utils/PasswordUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class PasswordUtilsTest {
+
+    @Test
+    void should_returnTrue_when_comparingTwoHashesOfSamePassword() throws Exception{
+        String password = "passwordTest";
```
Comment (VincentFleetwood) [inline user/src/test/java/com/jumar/user/utils/PasswordUtilsTest.java:11]: again, this is not testing business logic, really just a test of the hashing algorithm but we wouldn't write a test for that.

File: user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java (hunk 5 above)
Comment (VincentFleetwood) [reply user/src/main/java/com/jumar/user/services/impl/UserServiceImpl.java:40]: "Don't catch unless you know what to do with it"

File: user/src/main/java/com/jumar/user/controllers/UserController.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/")
+public class UserController {
+
+    private final UserService userService;
+
+    @PostMapping("users/create")
+    public ResponseEntity<User> createUser(@RequestBody CreateUserDto createUserDto) {
+        return new ResponseEntity<>(userService.createUser(createUserDto), HttpStatus.CREATED);
```
Comment (VincentFleetwood) [inline user/src/main/java/com/jumar/user/controllers/UserController.java:22]: if you want to do logging and handling of errors, do it here.

**Pull request merged**